artiq/applets/big_number.py

```diff
@@ -26,47 +27,77 @@ def __init__(self, args, ctl):
         QtWidgets.QStackedWidget.__init__(self)
         self.dataset_name = args.dataset
         self.ctl = ctl
+        self.value = None
+        self.metadata = {}
 
-        self.lcd_widget = QResponsiveLCDNumber()
-        self.lcd_widget.setDigitCount(args.digit_count)
-        self.lcd_widget.doubleClicked.connect(self.start_edit)
-        self.addWidget(self.lcd_widget)
+        self.display_widget = QResponsiveLabel()
+        self.display_widget.setAlignment(
+            QtCore.Qt.AlignRight | QtCore.Qt.AlignVCenter)
+        self.display_widget.doubleClicked.connect(self.start_edit)
+        self.addWidget(self.display_widget)
 
         self.edit_widget = QCancellableLineEdit()
-        self.edit_widget.setValidator(QtGui.QDoubleValidator())
-        self.edit_widget.setAlignment(QtCore.Qt.AlignRight)
+        self.edit_widget.setAlignment(
+            QtCore.Qt.AlignRight | QtCore.Qt.AlignVCenter)
         self.edit_widget.editCancelled.connect(self.cancel_edit)
         self.edit_widget.returnPressed.connect(self.confirm_edit)
         self.addWidget(self.edit_widget)
 
         font = QtGui.QFont()
         font.setPointSize(60)
+        self.display_widget.setFont(font)
         self.edit_widget.setFont(font)
 
-        self.setCurrentWidget(self.lcd_widget)
+        self.setCurrentWidget(self.display_widget)
+
+    def get_raw_format(self):
+        scale = scale_from_metadata(self.metadata)
+        value = self.value/scale
+        unit = self.metadata.get("unit", "")
+        return f"{value} {unit}".strip()
 
     def start_edit(self):
-        # QLCDNumber value property contains the value of zero
-        # if the displayed value is not a number.
-        self.edit_widget.setText(str(self.lcd_widget.value()))
+        if self.value is None:
+            self.edit_widget.setText("---")
+        else:
+            self.edit_widget.setText(self.get_raw_format())
         self.edit_widget.selectAll()
         self.edit_widget.setFocus()
         self.setCurrentWidget(self.edit_widget)
 
     def confirm_edit(self):
-        value = float(self.edit_widget.text())
-        self.ctl.set_dataset(self.dataset_name, value)
-        self.setCurrentWidget(self.lcd_widget)
+        text = self.edit_widget.text().split()
+        if not text:
+            self.cancel_edit()
+            return
+        try:
+            value = float(text[0])
+        except ValueError:
+            self.display_widget.setText("---")
+            self.setCurrentWidget(self.display_widget)
+            return
+        unit = text[1] if len(text) > 1 else ""
+        scale = scale_from_metadata({"unit": unit})
+        value = value*scale
+        precision = self.metadata.get("precision", None)
```

Comment:
None is redundant here since it is already the default. That being said, this is just a nitpick